mozilla_schema_generator/main_ping.py

```diff
@@ -31,17 +31,46 @@ def get_schema(self):
         return self._update_env(schema)
 
     def _update_env(self, schema):
-        user_pref_elem = {
+        integer = {"type": "integer"}
+        string = {"type": "string"}
+        string_map = {
+            "type": "object",
+            "additionalProperties": string,
+        }
+        nested_string_map = {
+            "type": "object",
+            "additionalProperties": string_map,
+        }
+        map_without_values = {
             "type": "object",
             "additionalProperties": {
-                "type": "string"
-            }
+                "type": "object",
+            },
         }
 
+        active_addons = prepend_properties(("environment", "addons", "activeAddons")) + ("additionalProperties", "properties")
+        simple_measurements = prepend_properties(("payload", "simpleMeasurements", ""))[:-1]
+        processes_dynamic = prepend_properties(("payload","processes", "dynamic", ""))[:-1]
+
+        schema.set_schema_elem(
+                prepend_properties(("environment", "settings", "userPrefs")), string_map)
+        schema.set_schema_elem(
+                prepend_properties(("environment", "system", "os", "version")), string)
         schema.set_schema_elem(
-                prepend_properties(("environment", "settings", "userPrefs")), user_pref_elem)
+                prepend_properties(("environment", "addons", "theme", "foreignInstall")), integer)
         schema.set_schema_elem(
-                prepend_properties(("environment", "system", "os", "version")), {"type": "string"})
+                prepend_properties(("payload", "addonDetails", "XPI")), map_without_values)
```

Comment:
adding XPI is blocked on [bug 1596600](https://bugzilla.mozilla.org/show_bug.cgi?id=1596600)

Reply:
There could add-ons may be XPI-based, if they're still being reported in the addons fields.

Reply:
> There could add-ons may be XPI-based, if they're still being reported in the addons fields.

did you mean to `s/There could//`?
yes, addons are either still XPI-based or still considered as such for the purposes of addonDetails

Reply:
I'm going to break out xpi into a separate PR